lib/Extractor/Solver.cpp

```diff
@@ -76,6 +76,164 @@ class BaseSolver : public Solver {
   BaseSolver(std::unique_ptr<SMTLIBSolver> SMTSolver, unsigned Timeout)
       : SMTSolver(std::move(SMTSolver)), Timeout(Timeout) {}
 
+  bool testZeroSign(const BlockPCs &BPCs,
+                    const std::vector<InstMapping> &PCs,
+                    APInt &Negative, Inst *LHS,
+                    InstContext &IC) {
+    unsigned W = LHS->Width;
+    Inst *Mask = IC.getConst(Negative);
+    InstMapping Mapping(IC.getInst(Inst::And, W, { LHS, Mask }), IC.getConst(APInt::getNullValue(W)));
+    bool IsSat;
+    std::error_code EC = SMTSolver->isSatisfiable(BuildQuery(IC, BPCs, PCs, Mapping, 0, /*Precondition=*/0),
+                                                  IsSat, 0, 0, Timeout);
+    if (EC)
+      llvm::report_fatal_error("stopping due to error");
+    return !IsSat;
+  }
+
+  bool testOneSign(const BlockPCs &BPCs,
+                   const std::vector<InstMapping> &PCs,
+                   APInt &Negative, Inst *LHS,
+                   InstContext &IC) {
+    unsigned W = LHS->Width;
+    Inst *Mask = IC.getConst(Negative);
+    InstMapping Mapping(IC.getInst(Inst::And, W, { LHS, Mask }), Mask);
+    bool IsSat;
+    std::error_code EC = SMTSolver->isSatisfiable(BuildQuery(IC, BPCs, PCs, Mapping, 0, /*Precondition=*/0),
+                                                  IsSat, 0, 0, Timeout);
+    if (EC)
+      llvm::report_fatal_error("stopping due to error");
+    return !IsSat;
+  }
+
+  std::error_code negative(const BlockPCs &BPCs,
+                           const std::vector<InstMapping> &PCs,
+                           Inst *LHS, APInt &Negative,
+                           InstContext &IC) override {
+    unsigned W = LHS->Width;
+    Negative = APInt::getNullValue(W);
+    APInt NegativeGuess = Negative | APInt::getOneBitSet(W, W-1);
+    if (testZeroSign(BPCs, PCs, NegativeGuess, LHS, IC))
+      Negative = APInt::getNullValue(W);
+    else if (testOneSign(BPCs, PCs, NegativeGuess, LHS, IC))
+      Negative = NegativeGuess;
+    return std::error_code();
+  }
+
+  std::error_code nonNegative(const BlockPCs &BPCs,
+                              const std::vector<InstMapping> &PCs,
+                              Inst *LHS, APInt &NonNegative,
+                              InstContext &IC) override {
+    unsigned W = LHS->Width;
+    NonNegative = APInt::getNullValue(W);
+    APInt NonNegativeGuess = NonNegative | APInt::getOneBitSet(W, W-1);
+    if (testZeroSign(BPCs, PCs, NonNegativeGuess, LHS, IC)) {
+      NonNegative = APInt::getNullValue(W);
+    } else if (testOneSign(BPCs, PCs, NonNegativeGuess, LHS, IC)) {
+      NonNegative = NonNegativeGuess;
+    } else {
+      NonNegative = NonNegativeGuess;
```

Comment:
it looks like this else branch and the previous "else if" branch contain the same code, that cannot be correct